lib/ui/painting/fragment_shader.cc

```diff
@@ -48,4 +49,78 @@ FragmentShader::FragmentShader(
 
 FragmentShader::~FragmentShader() = default;
 
+IMPLEMENT_WRAPPERTYPEINFO(ui, ReusableFragmentShader);
+
+ReusableFragmentShader::ReusableFragmentShader(
+    fml::RefPtr<FragmentProgram> program,
+    uint64_t float_count,
+    uint64_t sampler_count)
+    : program_(program),
+      uniform_data_(SkData::MakeUninitialized(
+          (float_count + 2 * sampler_count) * sizeof(float))),
+      samplers_(sampler_count),
+      float_count_(float_count) {}
+
+Dart_Handle ReusableFragmentShader::Create(Dart_Handle wrapper,
+                                           Dart_Handle program,
+                                           Dart_Handle float_count_handle,
+                                           Dart_Handle sampler_count_handle) {
+  auto* fragment_program =
+      tonic::DartConverter<FragmentProgram*>::FromDart(program);
+  uint64_t float_count =
+      tonic::DartConverter<uint64_t>::FromDart(float_count_handle);
+  uint64_t sampler_count =
+      tonic::DartConverter<uint64_t>::FromDart(sampler_count_handle);
+
+  auto res = fml::MakeRefCounted<ReusableFragmentShader>(
+      fml::Ref(fragment_program), float_count, sampler_count);
+  res->AssociateWithDartWrapper(wrapper);
+
+  void* raw_uniform_data =
+      reinterpret_cast<void*>(res->uniform_data_->writable_data());
+  return Dart_NewExternalTypedData(Dart_TypedData_kFloat32, raw_uniform_data,
+                                   float_count);
```

Comment:
I wonder if you pass in the typed data you can avoid it being `late final` in the Dart class. As in, it can be initialized to the right size in the dart ctor and passed in as an argument here and filled in on this side via `Dart_TypedDataAcquire`.

Reply:
I should have noted in the PR description, but one of the purposes of this PR is reducing/eliminating copies of the uniform data. If the uniform data is in the Dart heap, it would need to be copied out before the matching `Dart_TypedDataRelease`.

Reply:
Ahhh makes sense